flang/lib/Optimizer/CodeGen/Target.cpp

```diff
@@ -825,6 +825,51 @@ struct TargetAArch64 : public GenericTarget<TargetAArch64> {
     }
     return marshal;
   }
+
+  // Determine if the type is a Homogenous Floating-point Aggregate (HFA). An
+  // HFA is a record type with up to 4 floating-point members of the same type.
+  static bool isHFA(fir::RecordType ty) {
+    auto types = ty.getTypeList();
+    if (types.empty() || types.size() > 4) {
+      return false;
+    }
+
+    if (!isa_real(types.front().second)) {
+      return false;
+    }
```

Comment:
Derived types containing small fp arrays will be rejected here, is the following C struct an HFA? :

```
typedef struct {
    float x[2];
    float y;
} S;
```

Reply:
Ah. It is, yes. As is:
```
typedef struct {
  float x;
  float y;
} S1;

typedef struct {
  S1 s;
  float z;
} S2;
```
Good spot.